chroma-common/blockdevices/blockdevice_zfs.py

```diff
@@ -129,8 +124,19 @@ def lock_pool(self):
         while not self.lock.i_am_locking():
             try:
                 self.lock.acquire(self.LOCK_ACQUIRE_TIMEOUT)
+                with open(self.lock.lock_file, 'w') as f:
+                    f.writelines(str(self.lock.pid))
             except LockTimeout:
-                pass
+                # prune locks owned by non-existent processes
+                with open(self.lock.lock_file, 'r') as f:
+                    contents = f.readlines()
+
+                assert len(contents) == 1 and contents[0].isdigit(), \
+                    'unexpected contents of lockfile %s: %s' % (self.lock.lock_file, contents)
+
+                # validate pid holding lock exists
+                if not pid_exists(int(contents[0])):
```

Comment:
Which process is terminated?  The child that was called under the lock or the process doing the locking?

In the former case the parent should reap the terminated child and remove the lock.

In the latter, such a process should catch the signal that killed it and do the appropriate cleanup.  This should work for any signal except SIGKILL, however no process should ever send SIGKILL without first trying other less drastic signals such as SIGINT or SIGTERM, etc.[1]

Are we doing the above two things:

1. reaping children we started when they are killed unexpectedly and cleaning things up like lock files, etc. and
2. catching signals like SIGINT, SIGTERM, SIGHUP, etc. and doing the appropriate cleanup?

If we are, should lock breaking still be necessary and not be considered working around another bug?

[1] Sending a SIGKILL without first sending a signal that allows a process to catch it and clean itself up is anti-social at best, outright evil at worst and generally should be considered a bug that needs fixing.